### PR TITLE
[Multi-sig] feat: implement multi-sig support for funding an advanced payment

### LINKED ISSUE #3497: [Multi-sig] feat: implement multi-sig support for funding advanced payments
### Story

As a USER, I would like to be able to use multisig to fund an advanced payment.

### Description

With this, we should be able to support multisig for advanced payments, split payments and soon to be released streaming payments, since they all use the same contracts.

### Design

[Figma link](https://www.figma.com/design/SaO0Zj6e7yGqLzjEsxqVhy/Permissions?node-id=22101-46905&node-type=frame&t=88D3UBSl2i7HmUac-0)
image

### This issue is complete when...

- [ ] You can create a multisig motion for funding an expenditure
- [ ] The correct signees can vote on the motion
- [ ] Upon finalization, the funding step is applied
- [ ] It integrates into the existing flow seamlessly

Part of #3418

### PR DESCRIPTION
**Pull request by bassgeta**
## Description

This PR adds Multi-sig support for the funding step.
Note: for some reason, the payment builder widget can get stuck in a pending state, both in the confirm details and funding step... This was happening to me on `master` too, so we should definitely investigate why it happens, but I don't think we should fix it here...

[Block ingestor PR](https://github.com/JoinColony/block-ingestor/pull/290)
[Figma link](https://www.figma.com/design/SaO0Zj6e7yGqLzjEsxqVhy/Permissions?node-id=22101-46905&node-type=frame&t=88D3UBSl2i7HmUac-0)

## Testing

The testing itself isn't going to be a lot of work, but if you spot anything being broken, give me a shout.
Test it with `amy`, `fry` and `leela` with 3 browsers so we can cover the flow properly.

1. Install the Multi-sig extension, set the global threshold to 2
image
2. Give `leela` Multi-sig `Owner` permissions in `General` , `amy` the Multi-sig `Payer` in `General` and `fry` Multi-sig Payer in `Andromeda`
image
image
image
3. Create an advanced payment action sending out tokens to `amy` and `leela` from `General`
image
4. Get to the `Funding` step, choose Multi-sig as the decision method. Also verify that you get a notification
image
image
image
5. Refresh the page, verify that the widget loading isn't janky (just a skeleton should appear)
6. Let's try out rejection first, Remove your vote as `leela` and then Reject the multisig. Verify that the status pill changed to `Rejected` and that you get the funding button back. Verify that you get a rejected multisig notification.
image
image
image
7. Now create a multi sig motion for funding again. Verify that you can switch between the two.
image
8. Open up the payment action as `amy` and verify that she can sign it and you can switch between actions. The first one should be open by default
image
9. Open the action up as `fry` and verify that you can't vote on it, but you can peruse at your own leisure
image
10. Right back at it, as `amy` approve the funding step and Finalize the motion. Verify that you get notifications for approval and finalization
image
11. Now release as `leela`, wrap it up and verify that `amy` and `leela` got the funds
image
image
Right additionally let's see if `fry` can sign from `Andromeda`.
1. As `leela` create an advanced payment sending some CREDS to 1 recipient (we test an edge case where an advanced payment to 1 recipient is parsed as just `moveFunds` in the block-ingestor.
image
2. Get to the funding step and as `fry` choose Multi-sig as the decision method (just so we verify not only the payment creator can do that)
image
image
3. Remove your vote as `fry` and re-approve (just a sanity check)
4. As `leela` approve this motion, finalize it and see that you got the funds
image

### Additional tests
Since I've kinda rewrote the `getMoveFundsPermissionProofs` helper please also test
1. Move funds via permissions action still works
2. Funding an expenditure via permissions still works
3. Set the threshold to something like 10, try to fund an expenditure via Multi-sig and verify that this banner appears with the submit disabled
image
4. Verify that you can't fund an expenditure via multi-sig if you don't have the Funding multi-sig role in that team

#### Funding with multiple tokens:

Create and advanced payment with multiple tokens and try to fund via multi-sig.
Verify that the multi-sig was created with the proper `expenditureFunding` array by running this query:
```
getColonyAction(
    id: "the action id"
  ) {
    colonyId
    expenditureId
    expenditure {
      balances {
        amount
        tokenAddress
      }
    }
  }
```
image

Copy over the expenditureId and run this query
```
getMultiSigByExpenditureId(
    expenditureId: "copied expenditureId"
  ) {
    items {
      multiSigDomainId
      isExecuted
      isRejected
      expenditureFunding {
        amount
        tokenAddress
      }
    }
  }
```
image
Verify that both tokens are in the array. Finalize the funding multisig and rerun both queries (I just did it at once) and verify that the expenditure funding matches the multisig one:
image
image

## Diffs
I've noted the block-ingestor diffs in that PR.

**New stuff** ✨

* `fundExpenditureMultiSig` saga
* `MotionWidgetSkeleton` component
* `getMoveFundsActionDomain` helper which returns the domain in which the action should happen (if we are doing a motion, we pass that through and it overrides it)

**Changes** 🏗

* `PaymentBuilderWidget` now has multi-sig support
* `MultiSigWidget` can be outside the stepper, added a `variant` prop for that
* `FundingModal` can trigger funding via multi-sig
* added `expenditureFunding` to the `ColonyMultiSig` model
* `getMoveFundsPermissionProofs` now just returns the child skill indices required and realigned `moveFunds` and `fundExpenditure` sagas
* had to realign split payment and payment builder `getPayload` helpers to not always send through the createdIn domain (we really do need to split those payloads up depending on the decision method, but let's do it when we do multisig support for creating them)

Resolves #3497

### CODE REVIEW

**Review by iamsamgibbs — CHANGES_REQUESTED** [review]
Great work on this so far! These components seem like a real pain to deal with, so nice work getting this implemented so well!

I've found a few issues which I'll detail as best I can. My dev environment seemed to get more and more unstable the more testing I did (I think mostly related to this issue I just opened: #3673)

Multi-Sig option appears as expected:

Screenshot 2024-11-07 at 14 20 50

And does not appear when extension is not installed:

Screenshot 2024-11-07 at 15 06 51

Signatures component shows as expected:

Screenshot 2024-11-07 at 14 21 19

Screenshot 2024-11-07 at 14 22 29

I can approve, reject as either user without issue.

I did get stuck with this "Pending" button on multiple occasions though when confirming a rejection:
 
Screenshot 2024-11-07 at 14 22 49

https://github.com/user-attachments/assets/e54b5a3b-8a1a-4dcf-964e-967ad853cdea

I was unable to get a funding in a sub-domain to work though as either user:

As Fry:
Screenshot 2024-11-07 at 14 27 29

Screenshot 2024-11-07 at 14 29 11

As Leela:
Screenshot 2024-11-07 at 14 30 16

Notifications also aren't showing the correct information:
 
Screenshot 2024-11-07 at 14 31 00

I tested that your MoveFunds change on block-ingestor didn't break the notification there which still seems okay!

Screenshot 2024-11-07 at 14 31 53

Tested looking at failed motions / multi-sigs which works great:
https://github.com/user-attachments/assets/029970a7-e7ad-4e02-9b20-a8ed7b1ab294

There is an `Invalid threshold` message that is shown when the extension is uninstalled though. I didn't test this, but I suspect if the extension was re-installed, it would try and update based on the new extension settings. So this should be hidden if multi-sig being loaded has been uninstalled.
 
Screenshot 2024-11-07 at 15 28 07

There is no check on the modal that you actually have permissions to create the multi-sig (this is in Wayne where no user has multi-sig permissions at all)

Screenshot 2024-11-07 at 15 28 36

Also doesn't check if there are enough eligible signees:

Screenshot 2024-11-07 at 15 30 18

Appreciate there is a lot there! I'm not sure how much of this would make sense to handle in this issue or as separate issues. Nice job on getting it hooked up so well this far though. The actual core multi-sig aspect of it is all working nicely.

**Review by iamsamgibbs — COMMENTED** [review]
Nice job getting the multi-sig working in a sub-domain!

I'm now able to fund an expenditure in Andromeda as Fry with payer permissions only in Andromeda:

https://github.com/user-attachments/assets/b7bcf248-602b-4a71-a435-6dde128d3465

I had quite a few issues with the stepper not progressing to the next step. In the above video you can see it doesn't progress to review and seems like it would let me start another funding multi-sig / permissions action if I wanted.

Also found similar issue here where the multi-sig widget didn't load after starting funding:

https://github.com/user-attachments/assets/467c3ee3-e518-4120-9e52-6e882f3beb00

Also got stuck removing here:

Screenshot 2024-11-15 at 11 42 59

Also found that the multi-sig widget wasn't updating for Leela when Amy / Fry would approval and finalise.

I'll let someone else give it a review and see if it's just issues with my environment.

File: src/components/v5/common/ActionSidebar/partials/forms/PaymentBuilderForm/utils.ts [hunk 1]
```diff
@@ -18,15 +19,13 @@ export const getPaymentBuilderPayload = (
...
+): CreateExpenditurePayload | CreateStakedExpenditurePayload | null => {
   const colonyTokens = colony.tokens?.items.filter(notNull);
   const rootDomain = findDomainByNativeId(Id.RootDomain, colony);
   const createdInDomain =
-    findDomainByNativeId(values.createdIn, colony) || rootDomain;
-
-  if (!createdInDomain) {
-    return null;
-  }
+    values.decisionMethod === 'Permissions'
```
Comment (mmioana) [inline src/components/v5/common/ActionSidebar/partials/forms/PaymentBuilderForm/utils.ts:26]: Can we use here the following check?
```suggestion
    values.decisionMethod === DecisionMethod.Permissions
```

File: src/components/v5/common/ActionSidebar/partials/forms/SplitPaymentForm/utils.ts [hunk 2]
```diff
@@ -23,9 +24,13 @@ export const getSplitPaymentPayload = (
     team,
     description,
     payments,
+    decisionMethod,
   } = values;
   const rootDomain = findDomainByNativeId(Id.RootDomain, colony);
-  const createdInDomain = findDomainByNativeId(createdIn, colony) || rootDomain;
+  const createdInDomain =
+    decisionMethod === 'Permissions'
```
Comment (mmioana) [inline src/components/v5/common/ActionSidebar/partials/forms/SplitPaymentForm/utils.ts:31]: Also here
```suggestion
    decisionMethod === DecisionMethod.Permissions
```

**Review by mmioana — CHANGES_REQUESTED** [review]
Lots of sweat and tears must have been shed for this issue, but really awesome work @bassgeta 👏 

I'll start with the things that have been working when reviewing this PR and then will leave the points that still need work on 

`Installed the multi-sig extension`
Screenshot 2024-11-25 at 09 48 00

And gave `leela`, `amy` and `fry` the suggested permissions
Screenshot 2024-11-25 at 09 48 49
Screenshot 2024-11-25 at 09 49 16
Screenshot 2024-11-25 at 09 49 42

Created an advanced payment as `leela` and used `Multi-sig` for funding 
Screenshot 2024-11-25 at 09 51 02
Screenshot 2024-11-25 at 09 51 21

The `Multi-sig` funding notification is shown
Screenshot 2024-11-25 at 09 51 48

The proper skeleton is shown upon refresh
Screenshot 2024-11-25 at 09 52 17

Removed `leela`s approval
Screenshot 2024-11-25 at 09 52 37

Rejected as `leela`
Screenshot 2024-11-25 at 09 53 05
Screenshot 2024-11-25 at 09 53 10
Screenshot 2024-11-25 at 09 53 30
Then funded again the payment using `Multi-sig`
Screenshot 2024-11-25 at 09 54 00

Can switch between the current and previous funding actions

https://github.com/user-attachments/assets/639ca2ed-c812-4cfa-85df-a29b3f11fae6

`amy` can interact with the `Multi-sig` widget
Screenshot 2024-11-25 at 09 55 10

While `fry` is only in viewer-mode 👓 
Screenshot 2024-11-25 at 09 55 58

Approved as `amy`
Screenshot 2024-11-25 at 09 56 17
Screenshot 2024-11-25 at 09 56 33
Screenshot 2024-11-25 at 09 56 40

Payment has been released
Screenshot 2024-11-25 at 09 57 17
Screenshot 2024-11-25 at 09 57 34

Now creating and advanced payment in `Andromeda`
Screenshot 2024-11-25 at 09 59 05

`fry` can fund using `Multi-sig`
Screenshot 2024-11-25 at 10 00 17
`Approved` by default
Screenshot 2024-11-25 at 10 00 36

Remove `fry`'s approval
Screenshot 2024-11-25 at 10 00 46

And approve again
Screenshot 2024-11-25 at 10 01 03

Remove approval again and then reject 
Screenshot 2024-11-25 at 10 01 18
Screenshot 2024-11-25 at 10 01 38

Now moving some funds using `Permissions`
Screenshot 2024-11-25 at 10 04 21

Advanced payment using `Permissions` as funding
Screenshot 2024-11-25 at 10 05 14


Updating the `Multi-sig` threshold to a higher value than existing permission holders
Screenshot 2024-11-25 at 10 05 28
The error message is shown
Screenshot 2024-11-25 at 10 07 36

Trying to fund a payment in `Rocinate` as `fry`, the `Multi-sig` option is not displayed 💯 
Screenshot 2024-11-25 at 10 09 39

Now the issues part 😢 

## Issues

1. The `Manage permissions` completed action screen
It seems the wrong `Owner` permissions table is shown after assigning permissions to `amy` and `fry` - **however this should be treated in a separate issue**
Screenshot 2024-11-25 at 09 49 16
Screenshot 2024-11-25 at 09 49 23
Screenshot 2024-11-25 at 09 49 42
Screenshot 2024-11-25 at 09 49 47
2. Remove `fry` approval and then reject gets into an always pending state before page refresh, while `leela` can already see the `Fund` button
Screenshot 2024-11-25 at 10 01 18
Screenshot 2024-11-25 at 10 01 38
Screenshot 2024-11-25 at 10 03 50

3. While updating the `Multi-sig` threshold to a higher value than existing permission holders, the error box for the funding step is shown only after page refresh
Screenshot 2024-11-25 at 10 05 28
Screenshot 2024-11-25 at 10 06 53
Screenshot 2024-11-25 at 10 07 36

Nonetheless, amazing effort @bassgeta 🥇

**Review by mmioana — APPROVED** [review]
Awesome, 2nd point has been resolved with light-speed 💪 

Thanks for the fast change @bassgeta 
Screenshot 2024-11-25 at 11 21 03
Screenshot 2024-11-25 at 11 21 57
Screenshot 2024-11-25 at 11 22 06
Screenshot 2024-11-25 at 11 22 12
Screenshot 2024-11-25 at 11 22 20
Screenshot 2024-11-25 at 11 22 29

I will approve this PR and create a new issue for the 1st point 💯

**Review by mmioana — APPROVED** [review]
Stellar work @bassgeta 🥇 

Re-tested the main testing flow and it all worked as expected
Screenshot 2024-11-26 at 18 06 47
Screenshot 2024-11-26 at 18 07 16
Screenshot 2024-11-26 at 18 07 37
Screenshot 2024-11-26 at 18 08 04
Screenshot 2024-11-26 at 18 08 46
Screenshot 2024-11-26 at 18 09 03
Screenshot 2024-11-26 at 18 09 17
Screenshot 2024-11-26 at 18 09 36
Screenshot 2024-11-26 at 18 09 54
Screenshot 2024-11-26 at 18 11 11
Screenshot 2024-11-26 at 18 11 53
Screenshot 2024-11-26 at 18 12 08
Screenshot 2024-11-26 at 18 12 26
Screenshot 2024-11-26 at 18 13 31
Screenshot 2024-11-26 at 18 14 06
Screenshot 2024-11-26 at 18 15 21


Tested the `Advanced payment` funding for multiple tokens and it all looks good from my side 🥇 
Screenshot 2024-11-26 at 18 18 00
Screenshot 2024-11-26 at 18 18 18
Screenshot 2024-11-26 at 18 19 04
Screenshot 2024-11-26 at 18 19 44
Screenshot 2024-11-26 at 18 20 41

Thank you for the huge effort on this issue 💯 let's get it merged!

**Review by jakubcolony — APPROVED** [review]
Can confirm `expenditureFunding` is now showing all the tokens being funded 👍 

image

Nice one, let's get this merged! 💯

**Review by iamsamgibbs — APPROVED** [review]
Nice job handling the case with multiple tokens. Tested and all working as expected!

Screenshot 2024-11-27 at 11 33 41

Screenshot 2024-11-27 at 11 34 56
